cdap-client/src/main/java/io/cdap/cdap/client/ApplicationClient.java

```diff
@@ -239,8 +240,31 @@ public ApplicationDetail get(ApplicationId appId)
    * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
    */
   public void upgradeApplication(ApplicationId app)
+    throws ApplicationNotFoundException, IOException, UnauthenticatedException {
+    upgradeApplication(app, Collections.EMPTY_SET, false);
+  }
+
+  /**
+   * Upgrades an application.
+   *
+   * @param app the application to delete
+   * @param artifactScopes artifact scopes to consider while searching for latest artifacts.
+   * @param allowSnapshot should consider snapshot artifacts or not.
+   * @throws ApplicationNotFoundException if the application with the given ID was not found
+   * @throws IOException if a network error occurred
+   * @throws UnauthenticatedException if the request is not authorized successfully in the gateway server
+   */
+  public void upgradeApplication(ApplicationId app, Set<String> artifactScopes, boolean allowSnapshot)
     throws ApplicationNotFoundException, IOException, UnauthenticatedException, UnauthorizedException {
-    String path = String.format("apps/%s/upgrade", app.getApplication());
+    String path = String.format("apps/%s/upgrade?allowSnapshot=%s", app.getApplication(),
+                                Boolean.toString(allowSnapshot));
+
+    if (artifactScopes != null && !artifactScopes.isEmpty()) {
```

Comment:
artifactScopes won't be null, right? Meaning you don't need the if check at all. Just for-looping it is ok.